### PR TITLE
Tree conv op

### LINKED ISSUE #14396: Add Tree-Based Convolution Operator
The paper of Tree-Based Convolutioin Operator is here:
https://arxiv.org/abs/1409.5718v1

### PR DESCRIPTION
**Pull request by zzhzz**
resolved #14396

### CODE REVIEW

File: paddle/fluid/operators/math/tree2col.cu [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stack>
+#include "paddle/fluid/operators/math/math_function.h"
+#include "paddle/fluid/operators/math/tree2col.h"
+
+namespace paddle {
+using Tensor = framework::Tensor;
+using Node = paddle::operators::math::TreeNode;
```
Comment (heavengate) [inline paddle/fluid/operators/math/tree2col.cu]: It is recommended to put this in namespace math, for that paddle is the outer namespace, putting it here is in risk to incur conflicts with other ops.

File: paddle/fluid/operators/math/tree2col.cc [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+              eta_r * features[id * feature_size + i];
+          patch_data[pointer_base + i * 3 + 2] +=
+              eta_t * features[id * feature_size + i];
+        }
+      }
+      patch_count++;
+    }
+    patch->Resize(
+        framework::make_ddim({static_cast<int64_t>(patch_count),
+                              static_cast<int64_t>(patch_elem_size)}));
```
Comment (heavengate) [inline paddle/fluid/operators/math/tree2col.cc]: framework::make_ddim is not necessary for the first paramter, you can simply use {static_cast<int64_t(patch_counr), static_cast<int64_t>(patch_elem_size)} here, and will make your code more clearly, and same in mutable_data api.

File: paddle/fluid/operators/math/tree2col.cc [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  }
+};
```
Comment (heavengate) [inline paddle/fluid/operators/math/tree2col.cc:133]: It is recommended to use same blank line to separate these code.

File: paddle/fluid/operators/tree_conv_op.cc [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasOutput("Out"));
+    auto edge_dims = ctx->GetInputDim("EdgeSet");
+    auto vector_dims = ctx->GetInputDim("NodesVector");
+    auto filter_dims = ctx->GetInputDim("Filter");
+    PADDLE_ENFORCE_EQ(edge_dims.size(), 3);
+    PADDLE_ENFORCE_EQ(vector_dims.size(), 3);
+    PADDLE_ENFORCE_EQ(filter_dims.size(), 4);
+    PADDLE_ENFORCE_EQ(filter_dims[1], 3);
+    PADDLE_ENFORCE_EQ(filter_dims[0], vector_dims[2]);
```
Comment (heavengate) [inline paddle/fluid/operators/tree_conv_op.cc]: try to add some error messages in the 3rd parameter.

File: paddle/fluid/operators/math/tree2col.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <array>
+#include <unordered_map>
+#include <vector>
+#include "paddle/fluid/framework/tensor.h"
+#include "paddle/fluid/operators/math/blas.h"
```
Comment (heavengate) [inline paddle/fluid/operators/math/tree2col.h]: it seem blas.h was not used in this head file, if it is, it is recommended to move it to .cc or .cu file which used it, as for head file is the file to indicate APIs, try to make it succinct.

**Review by heavengate — COMMENTED** [review]
Could you please add python API in Paddle/python/paddle/fluid/layers/nn.py and add python API test in Paddle/python/paddle/fluid/tests/unittests/test_layers.py.

**Comment by luotao1** [comment]
- please reply to each comment by reviewer: http://www.paddlepaddle.org/documentation/docs/zh/1.1/advanced_usage/development/contribute_to_paddle.html#permalink-14--
- you can add `test=develop` in commit to trigger the CI

File: paddle/fluid/operators/math/tree2col.cu (hunk 1 above)
Comment (zzhzz) [reply paddle/fluid/operators/math/tree2col.cu]: Done

File: paddle/fluid/operators/math/tree2col.cc (hunk 2 above)
Comment (zzhzz) [reply paddle/fluid/operators/math/tree2col.cc]: Done

File: paddle/fluid/operators/math/tree2col.cc (hunk 3 above)
Comment (zzhzz) [reply paddle/fluid/operators/math/tree2col.cc:133]: Done

File: paddle/fluid/operators/math/tree2col.h (hunk 5 above)
Comment (zzhzz) [reply paddle/fluid/operators/math/tree2col.h]: Done

File: paddle/fluid/operators/tree_conv_op.cc (hunk 4 above)
Comment (zzhzz) [reply paddle/fluid/operators/tree_conv_op.cc]: Done

**Comment by zzhzz** [comment]
I change the python API and API.spec, the PR_CI gave me an error: 
```
[20:59:32][Step 1/1] current pr 14404 got approvals: FALSE
[20:59:32][Step 1/1] You must have at least 2 approvals for the api change!
[20:59:44][Step 1/1] Process exited with code 1
[20:59:44][Step 1/1] Process exited with code 1
[20:59:44][Step 1/1] Step Build and test (Command Line) failed
```
please help me fix this, thanks for your patience. @heavengate

**Comment by heavengate** [comment]
API changes should have 2 approves from membesr who have API approve permission，try to ask for reviews from @qingqing01 @panyx0718 @baiyfbupt

File: paddle/fluid/operators/tree_conv_op.cc [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasOutput("Out"));
+    auto edge_dims = ctx->GetInputDim("EdgeSet");
+    auto vector_dims = ctx->GetInputDim("NodesVector");
+    auto filter_dims = ctx->GetInputDim("Filter");
+    PADDLE_ENFORCE_EQ(
+        edge_dims[2], 2,
+        "The shape of EdgeSet Tensor is [batch_size, max_tree_node_size, 2]");
+    PADDLE_ENFORCE_EQ(edge_dims.size(), 3,
+                      "The dimension of EdgeSet Tensor is 3");
```
Comment (heavengate) [inline paddle/fluid/operators/tree_conv_op.cc]: is -> should be

File: paddle/fluid/operators/tree_conv_op.cc [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    PADDLE_ENFORCE_EQ(vector_dims.size(), 3, "The dimension of Input Tensor");
```
Comment (heavengate) [inline paddle/fluid/operators/tree_conv_op.cc]: message imcomplete, please fix it

File: paddle/fluid/operators/tree_conv_op.cc [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasOutput("Out"));
+    auto edge_dims = ctx->GetInputDim("EdgeSet");
+    auto vector_dims = ctx->GetInputDim("NodesVector");
+    auto filter_dims = ctx->GetInputDim("Filter");
+    PADDLE_ENFORCE_EQ(
+        edge_dims[2], 2,
+        "The shape of EdgeSet Tensor is [batch_size, max_tree_node_size, 2]");
```
Comment (heavengate) [inline paddle/fluid/operators/tree_conv_op.cc]: "Input(EdgeSet) dim[2] should be 2"       some like this

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -8425,3 +8426,52 @@ def bilinear_tensor_product(x,
...
+              param_attr=None,
+              bias_attr=None):
+    """
+    ${comment}
+
+    Args:
+        nodes_vector(${nodes_vector_type}): ${nodes_vector_comment}
+        edge_set(${edge_set_type}): ${edge_set_comment}
+        output_size(int): output feature width
+        num_filters(int, default 1): number of filters
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py]: For doc style consistency， we generally make it like "num_filters(int): number of filters, default 1"

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -8425,3 +8426,52 @@ def bilinear_tensor_product(x,
...
+        max_depth(int, default 2): max depth of filters
+        act(str, default tanh): activation function
+        param_attr(ParamAttr, default None): the parameter attribute for the filters
+        bias_attr(ParamAttr, default None): the parameter attribute for the bias of this layer
+    Returns:
+        out(${out_type}): ${out_comment}
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py:8581]: try to give python code example

**Review by heavengate — COMMENTED** [review]
Good job!

File: paddle/fluid/operators/tree_conv_op.cc (hunk 6 above)
Comment (zzhzz) [reply paddle/fluid/operators/tree_conv_op.cc]: all done

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -8472,3 +8473,74 @@ def bilinear_tensor_product(x,
...
+        param_attr(ParamAttr): the parameter attribute for the filters, default None
+        bias_attr(ParamAttr): the parameter attribute for the bias of this layer, default None
+
+    Returns:
+        out(${out_type}): ${out_comment}
+
+    Examples:
+        .. code-block:: python
+          nodes_vector = layers.data(name='vectors', shape=[None, 10, 5], dtype='float32)
+          # nodes_vector tensor
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: No need for `#`?

**Review by panyx0718 — COMMENTED** [review]
interesting. who is requesting this operator?

File: python/paddle/fluid/layers/nn.py (hunk 11 above)
Comment (zzhzz) [reply python/paddle/fluid/layers/nn.py]: I'll fix my code with new Memory mechanism and commit later, and I'll remove the comment sign .

Comment (luotao1) [reply python/paddle/fluid/layers/nn.py]: please add a blank line between line8505-8506 to polish annotation correctly. @shanyi15 could you check the annotation of this API?

**Comment by zzhzz** [comment]
I'll start a new pull request about this operator @luotao1 , and I'll close this pull request

**Pull request closed without merging**